### PR TITLE
Fix bug of generate-unmatched-brackets in CodeGenC::PrintSSAAssign

### PR DESCRIPTION
**Pull request by Light-of-Hers**
Fix the bug discussed in https://github.com/apache/incubator-tvm/issues/6884 @tqchen

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks @Light-of-Hers , can you enhance the code to still remove extra brackets if they match?

The idea would be to add a CheckBracketMatch code that scans the string and use counter to check if the two bracket indeed matches each other (and remove if they do)

**Comment by Light-of-Hers** [comment]
Updated @tqchen 
Maybe future we can selectively generate surrouding brackets of sub-expressions based the operators' associativity/priority and code-gen context.

File: src/target/source/codegen_c.cc [hunk 1]
```diff
@@ -139,11 +139,29 @@ void CodeGenC::PrintExpr(const PrimExpr& n, std::ostream& os) {  // NOLINT(*)
   }
 }
 
+static bool CheckOutmostBracketMatch(const std::string& s) {
+  if (s.front() == '(' && s.back() == ')') {
```
Comment (tqchen) [inline src/target/source/codegen_c.cc:143]: need tyo first check the length to avoid overflow

File: src/target/source/codegen_c.cc [hunk 2]
```diff
@@ -139,11 +139,29 @@ void CodeGenC::PrintExpr(const PrimExpr& n, std::ostream& os) {  // NOLINT(*)
...
+    int len = s.size();
```
Comment (tqchen) [inline src/target/source/codegen_c.cc:144]: size_t

File: src/target/source/codegen_c.cc [hunk 3]
```diff
@@ -139,11 +139,29 @@ void CodeGenC::PrintExpr(const PrimExpr& n, std::ostream& os) {  // NOLINT(*)
...
+  return false;
+}
+
 void CodeGenC::PrintSSAAssign(const std::string& target, const std::string& src, DataType t) {
   PrintType(t, stream);
   stream << ' ' << target << " = ";
-  if (src.length() > 3 && src[0] == '(' && src[src.length() - 1] == ')') {
-    stream << src.substr(1, src.length() - 2);
+  if (CheckOutmostBracketMatch(src)) {
+    stream << src.substr(1, src.length() - 1);
```
Comment (tqchen) [inline src/target/source/codegen_c.cc:164]: change this line back as we need to also remove the bracket in the end

File: src/target/source/codegen_c.cc [hunk 4]
```diff
@@ -139,11 +139,29 @@ void CodeGenC::PrintExpr(const PrimExpr& n, std::ostream& os) {  // NOLINT(*)
   }
 }
 
+static bool CheckOutmostBracketMatch(const std::string& s) {
```
Comment (tqchen) [inline src/target/source/codegen_c.cc:142]: CheckOutermostBracketMatch

Comment (tqchen) [reply src/target/source/codegen_c.cc:142]: Move static after all impl of the non-static function

**Review by tqchen — CHANGES_REQUESTED** [review]

**Comment by Light-of-Hers** [comment]
Sorry, some low-level mistakes😂. Is it okay now? @tqchen

**Review by tqchen — APPROVED** [review]

**Comment by tqchen** [comment]
Thanks @Light-of-Hers !

**Pull request merged**